lib/driver.js

```diff
@@ -671,9 +666,23 @@ class XCUITestDriver extends BaseDriver {
   }
 
   async installToSimulator () {
-    log.debug(`Installing app '${this.opts.app}' on device`);
-    if (this.opts.fullReset && this.opts.bundleId) {
-      log.debug('Full reset requested. Removing app from device');
+    if (!this.opts.app) {
+      log.debug('No app path is given. Nothing to install.');
+      return;
+    }
+
+    log.debug(`Installing app '${this.opts.app}' on device...`);
+    if (this.opts.bundleId) {
+      if (this.opts.noReset === true) {
+        // TODO: Replace this with getAppContainer verification from node-simctl package
+        // after Xcode 7 and older are deprecated
+        const appDirs = await this.opts.device.getAppDirs(path.basename(this.opts.app), this.opts.bundleId);
+        if (appDirs.length > 0) {
+          log.debug(`App ${this.opts.bundleId} seems to be already installed. No need to reinstall.`);
+          return;
+        }
+      }
+      log.debug('Reset requested. Removing app from the device');
```

Comment:
We will get to this point if `noReset` is `true` but the app is not installed. Also we will get here if `noReset` is `false` _and_ `fullReset` is false. Neither seems right.